### PR TITLE
[Pipeline] AnimateDiff SDXL

### LINKED ISSUE #6158: Feature Request: add SDXL support to animatediff pipeline
**Is your feature request related to a problem? Please describe.**
Currently the AnimateDiff pipeline only supports SD 1.4 and SD 1.5. AnimateDiff with SDXL seems like a great addition as the quality is greater.

**Describe the solution you'd like.**
Currently the MotionAdapter cannot load the SDXL motion module : 

The following results in an error : 
```
adapter = MotionAdapter.from_pretrained("vladmandic/animatediff-sdxl", torch_dtype=torch.float16)
```

```
ValueError: Cannot load <class 'diffusers.models.unet_motion_model.MotionAdapter'> from vladmandic/animatediff-sdxl because the following keys are missing: 
 up_blocks.1.motion_modules.0.norm.bias, down_blocks.2.motion_modules.0.proj_out.bias, up_blocks.1.motion_modules.1.transformer_blocks.0.ff.net.2.bias, down_blocks.3.motion_modules.1.transformer_blocks.0.ff.net.0.proj.weight, 
...
```

**Describe alternatives you've considered.**
An alternative solution would be to take inspirtaion from official implementation at https://github.com/guoyww/AnimateDiff
or from https://github.com/s9roll7/animatediff-cli-prompt-travel/tree/main/src/animatediff 

**Additional context.**
Official SDXL motion module : https://huggingface.co/guoyww/animatediff/blob/main/mm_sdxl_v10_beta.ckpt

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
# What does this PR do?

Fixes #6158.

Attempt at integrating https://github.com/guoyww/AnimateDiff/tree/sdxl.

Relevant discussion: https://github.com/huggingface/diffusers/pull/5928#issuecomment-1851737456

Continuation of #6195. Unfortunately, the code base from current main and the one in 6195 felt really divergent. I ended up messing my local branch. I thought starting fresh is a better idea since we had many unwanted changes in the original PR. This is still a WIP.

Thanks to [ModelsLab](https://modelslab.com/) for providing GPU support for faster testing.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@DN6 @sayakpaul @guoyww

### CODE REVIEW

**Comment by a-r-r-o-w** [comment]
Link to [Colab](https://drive.google.com/file/d/1076fX0AtMgYNdQyYF-mwug8Bl2SfZ8Be/view?usp=sharing).

<details>
<summary> Target usage </summary>

```py
import torch
from diffusers import AnimateDiffPipeline, DDIMScheduler, EulerDiscreteScheduler, DEISMultistepScheduler
from diffusers.models import MotionAdapter
from diffusers import AnimateDiffSDXLPipeline

adapter = MotionAdapter.from_pretrained("a-r-r-o-w/animatediff-motion-adapter-sdxl-beta", torch_dtype=torch.float16)

# model_id = "stabilityai/stable-diffusion-xl-base-1.0"
model_id = "stablediffusionapi/dynavision-xl-v0610"
# model_id = "Lykon/dreamshaper-xl-1-0"

# scheduler = EulerDiscreteScheduler.from_pretrained(
# scheduler = DEISMultistepScheduler.from_pretrained(
scheduler = DDIMScheduler.from_pretrained(
    model_id,
    subfolder="scheduler",
    clip_sample=False,
    timestep_spacing="linspace",
    beta_schedule="linear",
    steps_offset=1,
)
pipe = AnimateDiffSDXLPipeline.from_pretrained(
    model_id,
    motion_adapter=adapter,
    scheduler=scheduler,
    torch_dtype=torch.float16,
    variant="fp16",
).to("cuda")

pipe.enable_vae_slicing()
pipe.enable_vae_tiling()

result = pipe(
    prompt="a panda surfing in the ocean, realistic, hyperrealism, high quality",
    negative_prompt="low quality, worst quality",
    num_inference_steps=20,
    guidance_scale=8,
    width=1024,
    height=1024,
    num_frames=16,
)

from diffusers.utils import export_to_gif
export_to_gif(result.frames[0], "animation.gif")
```
</details>

<details>
<summary> Initial results (manually downscaled due to max upload size) </summary>

<table>
<tr>
  <td colspan="3" align=center><strong>SDXL Base</strong></td>
</tr>
<tr>
  <td align=center></td>
  <td align=center></td>
</tr>
<tr>
  <td colspan="3" align=center><strong>DynaVisionXL</strong></td>
</tr>
<tr>
  <td align=center></td>
  <td align=center></td>
  <td align=center></td>
</tr>
<tr>
  <td colspan="3" align=center><strong>DreamshaperXL</strong></td>
</tr>
<tr>
  <td align=center></td>
  <td align=center></td>
</tr>
</table>

</details>

Still experimenting and trying to find if I missed something. Quality seems to be okay-ish but definitely need to look for better parameters/models. SDXL checkpoint by guoyww is still a beta release, and maybe we could wait for it before considering a merge?

File: src/diffusers/models/unets/unet_motion_model.py
Comment (DN6) [inline src/diffusers/models/unets/unet_motion_model.py:227]: We can hardcode these values right? They don't need to be arguments?

File: src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py
Comment (DN6) [inline src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py:913]: Repeated

File: src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py
Comment (DN6) [inline src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py:1279]: Did we want to enable free init from the start? Or add after merging this in?

File: src/diffusers/models/unets/unet_motion_model.py
Comment (DN6) [inline src/diffusers/models/unets/unet_motion_model.py:229]: I don't think this argument is used anywhere? Can we remove it?

File: src/diffusers/models/unets/unet_motion_model.py
Comment (DN6) [inline src/diffusers/models/unets/unet_motion_model.py:470]: I think we might need to leave this as is unfortunately.  `attention_head_dim` for SD based configs are incorrectly named, because they actually set the number of attention heads. 

The Motion Model can only really work with configs where "num_attention_heads" is set to `None`. We're working on fixing the issue and we will try to see if we can deprecate the use of `attention_head_dim`.

**Review by DN6 — COMMENTED** [review]
Looking good 👍🏽 Left a few comments.

File: src/diffusers/models/unets/unet_motion_model.py
Comment (a-r-r-o-w) [reply src/diffusers/models/unets/unet_motion_model.py:194]: I added these due to the warnings that show up when the config does not match the init parameters. I will revert these changes.

File: src/diffusers/models/unets/unet_motion_model.py
Comment (a-r-r-o-w) [reply src/diffusers/models/unets/unet_motion_model.py:227]: I believe conv_in_kernel and conv_out_kernel are always 3, or atleast I haven't seen anything different so we can revert back to how it was before. For the rest, I think they are needed for correct initialization from existing SDXL configs.

**Comment by a-r-r-o-w** [comment]
It'd be great to support the SDXL version but since there are no official checkpoints or a training script, it makes it harder for the community to experiment with and does not seem like a good feature to add in core diffusers. I will try cooking a motion adapter training script based on what I understand so far when I find the time, and try reaching out to the authors over other mediums.

cc @guoyww @limbo0000 @AnyiRao @wyhsirius Thank you for your amazing work! Any suggestions and updates would be awesome :heart:

File: src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py:913]: These are different, no? guidance_scale and guidance_rescale

File: src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py:1279]: Let's take it up here since FreeInit has already been well integrated

File: src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py
Comment (DN6) [reply src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py:935]: Whoops. Mistake. Ignore then.

File: src/diffusers/models/unets/unet_motion_model.py
Comment (DN6) [reply src/diffusers/models/unets/unet_motion_model.py:468]: It would be handled correctly right? The check for `num_attention_heads` returns a None and then it gets set to the `attention_head_dim`? Seems to work here

```python
from diffusers import MotionAdapter, AnimateDiffSDXLPipeline

adapter = MotionAdapter.from_pretrained("a-r-r-o-w/animatediff-motion-adapter-sdxl-beta")
pipe = AnimateDiffSDXLPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", motion_adapter=adapter)
print(pipe.unet.config.num_attention_heads)
# output is: [5, 10, 20]
```

Comment (a-r-r-o-w) [reply src/diffusers/models/unets/unet_motion_model.py:468]: Oh, sorry! I think I added this with the thought that if we're initializing using `config["attention_head_dim"]`, it must exist as a key in the config to not have KeyError thrown.

You're right, I will revert these changes soon

**Comment by yiyixuxu** [comment]
@a-r-r-o-w @DN6 
does that mean we want to add to the community folder for now? what's the plan?

**Comment by a-r-r-o-w** [comment]
> @a-r-r-o-w @DN6 does that mean we want to add to the community folder for now? what's the plan?

It's a bit of a difficult situation since no official checkpoints exist. The pipeline can be moved into community but the modelling changes here are more general and add functionality without breaking support for existing motion adapters (I hope, since I'm yet to test properly. The currently failing tests are only specific to what I added which is a positive), so I think that should be okay. The changes to the UNetMotionModel essentially just replicate the logic of UNet2dConditionModel i.e. there's nothing new in modelling parts as such. Let me know your thoughts and I'll adapt the code accordingly, thanks.

**Comment by DN6** [comment]
> @a-r-r-o-w @DN6 does that mean we want to add to the community folder for now? what's the plan?

We are seeing some traction on the Motion Adapters. So I it would be really nice to have this in our core pipelines. Unfortunately, as @a-r-r-o-w mentioned, the checkpoint is still technically in beta, so perhaps community is the best place for moment.

**Comment by sayakpaul** [comment]
Cc: @DN6 could you share any progress on https://github.com/huggingface/diffusers/pull/6721#issuecomment-1969147261? And this is definitely not stale I think?

File: src/diffusers/models/unets/unet_motion_model.py
Comment (a-r-r-o-w) [reply src/diffusers/models/unets/unet_motion_model.py:195]: @DN6 Is it true that for all SDXL models, these parameters are True and 0 respectively? I haven't looked into it much but from the few models I checked, it seems like so. I will hardcode if you give it a final thumbs up.

**Comment by a-r-r-o-w** [comment]
Sharing a result with FreeInit enabled for same conditions with Dreamshaper XL 1.0 after the latest changes:

<table>
<tr>
<td align=center> without freeinit </td>
<td align=center> with freeinit </td>
</tr>
<tr>
<td align=center></td>
<td align=center></td>
</tr>
</table>

File: src/diffusers/models/unets/unet_motion_model.py
Comment (DN6) [reply src/diffusers/models/unets/unet_motion_model.py:195]: Hi @a-r-r-o-w Yes these should be True and 0 for all SDXL checkpoints.

File: src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py
Comment (DN6) [inline src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py:299]: Shouldn't this be handled inside `from_unet2d`?

File: src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py
Comment (DN6) [inline src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py:301]: Can we change this to 
https://github.com/huggingface/diffusers/blob/699dfb084ca12be9c1cd34ad7b6d0fffb9f0ebff/src/diffusers/pipelines/pia/pipeline_pia.py#L206

**Review by DN6 — COMMENTED** [review]
Looks good! Just a couple of minor things to take care of.

File: src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/animatediff/pipeline_animatediff_sdxl.py]: Indeed, they are already handled here: https://github.com/huggingface/diffusers/blob/2b618eaf8c852464a48ce800cd9e55ccc2ba6d2a/src/diffusers/models/unets/unet_motion_model.py#L440

Seems like forgot cleaning it up here; done now thanks!

**Comment by a-r-r-o-w** [comment]
Just tested SD1.5 models with the changes to UNetMotionModel. Everything seems to be working for both SD and SDXL. So I think this is very close to completion. @DN6 The sdxl motion adapter checkpoint in diffusers format is available on my account with about ~500 downloads so far. Will you be able to move it to the authors' accounts or are we okay with this?

**Comment by DN6** [comment]
@a-r-r-o-w I think it's fine to have it on your account since you worked on the conversion.

**Comment by sayakpaul** [comment]
@DN6 the checkpoints are better in their original organizations actually. The model card can contain a note about Aryan's contributions but they should ideally reside under the original org/author.

**Comment by DN6** [comment]
> @DN6 the checkpoints are better in their original organizations actually. The model card can contain a note about Aryan's contributions but they should ideally reside under the original org/author.

Cool in that case we'll try to get in touch with the authors and move the checkpoints @a-r-r-o-w. We can cite your work in the model card.

**Comment by a-r-r-o-w** [comment]
Sounds good to me. It was only a few line changes to the already existing motion adapter script and not much work so it doesn't really matter :)

> FAILED tests/pipelines/animatediff/test_animatediff_sdxl.py::AnimateDiffPipelineSDXLFastTests::test_save_load_optional_components - AttributeError: 'NoneType' object has no attribute 'tokenize'

All failing tests fixed with the latest commit locally. Seems like the above was handled differently for SDXL so I've copied over the logic from SDXL tests.

**Comment by a-r-r-o-w** [comment]
> Cool in that case we'll try to get in touch with the authors and move the checkpoints @a-r-r-o-w. We can cite your work in the model card.

Any updates from them @DN6?

**Comment by DN6** [comment]
Hi @a-r-r-o-w Yes we'll move it. 

We can just transfer the checkpoint from your org to theirs. Is everything ready on your end?

Could you prep the model card and add something along the lines of "Converted to Diffusers by  @a-r-r-o-w"  You can link it to your preferred profile (GitHub or the Hub) so that your work is also attributed.

**Comment by a-r-r-o-w** [comment]
@DN6 Hey! I believe everything should be good for transfer hopefully. Please find the checkpoint [here](https://huggingface.co/a-r-r-o-w/animatediff-motion-adapter-sdxl-beta).

File: src/diffusers/models/unets/unet_motion_model.py
Comment (daz-synth) [inline src/diffusers/models/unets/unet_motion_model.py:285]: Could you also load add_embedding in from_unet2d? Something like:

```
if hasattr(model, "add_embedding"):
    model.add_embedding.load_state_dict(unet.add_embedding.state_dict())
```

File: docs/source/en/api/pipelines/animatediff.md
Comment (DN6) [inline docs/source/en/api/pipelines/animatediff.md:115]:
```suggestion
adapter = MotionAdapter.from_pretrained("
guoyww/animatediff-motion-adapter-sdxl-beta", torch_dtype=torch.float16)
```

**Comment by DN6** [comment]
@a-r-r-o-w Just opened a PR to update the code snippet on the model card in the docs. I think we're ready to go once those changes are made. 👍🏽

File: src/diffusers/models/unets/unet_motion_model.py
Comment (a-r-r-o-w) [reply src/diffusers/models/unets/unet_motion_model.py:285]: If we take a look at current unet_2d_condition.py modelling code, the team has refactored out these changes into separate helper functions. I think that because unet_motion_model.py is mostly a copy of that, we can adapt those changes here and therefore all the functionality one would need. We can take it up in a future PR in my opinion (also I'm afraid I will not have time to test things thoroughly).

**Review by DN6 — APPROVED** [review]
Appreciate the patience with this @a-r-r-o-w. Good to merge 👍🏽

**Pull request merged**
